app/abci/utils.go

```diff
@@ -0,0 +1,232 @@
+	if ctx.BlockHeight() <= 1 {
+		return false
+	}
+
+	return cp.Abci.VoteExtensionsEnableHeight < ctx.BlockHeight()
+}
+
+// ValidatorStore defines the interface contract require for verifying vote
+// extension signatures. Typically, this will be implemented by the x/staking
+// module, which has knowledge of the CometBFT public key.
+type ValidatorStore interface {
+	GetPubKeyByConsAddr(context.Context, sdk.ConsAddress) (cmtprotocrypto.PublicKey, error)
+}
+
+// ValidateVoteExtensions defines a helper function for verifying vote extension
+// signatures that may be passed or manually injected into a block proposal from
+// a proposer in PrepareProposal. It returns an error if any signature is invalid
+// or if unexpected vote extensions and/or signatures are found or less than 2/3
+// power is received.
+func ValidateVoteExtensions(
+	ctx sdk.Context,
+	valStore ValidatorStore,
+	extCommit cometabci.ExtendedCommitInfo,
+) error {
+	currentHeight := ctx.HeaderInfo().Height
+	chainID := ctx.HeaderInfo().ChainID
+	commitInfo := ctx.CometInfo().GetLastCommit()
+
+	// Check that both extCommit + commit are ordered in accordance with vp/address.
+	if err := ValidateExtendedCommitAgainstLastCommit(extCommit, commitInfo); err != nil {
+		return err
+	}
+
+	// Start checking vote extensions only **after** the vote extensions enable
+	// height, because when `currentHeight == VoteExtensionsEnableHeight`
+	// PrepareProposal doesn't get any vote extensions in its request.
+	extensionsEnabled := VoteExtensionsEnabled(ctx)
+	marshalDelimitedFn := func(msg proto.Message) ([]byte, error) {
+		var buf bytes.Buffer
+		if err := protoio.NewDelimitedWriter(&buf).WriteMsg(msg); err != nil {
+			return nil, err
+		}
+
+		return buf.Bytes(), nil
+	}
+
+	var (
+		// Total voting power of all vote extensions.
+		totalVP int64
+		// Total voting power of all validators that submitted valid vote extensions.
+		sumVP int64
+	)
+
+	for _, vote := range extCommit.Votes {
+		totalVP += vote.Validator.Power
+
+		if extensionsEnabled {
+			if vote.BlockIdFlag == cmtproto.BlockIDFlagCommit && len(vote.ExtensionSignature) == 0 {
+				return fmt.Errorf("vote extension signature is missing; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+			if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit && len(vote.VoteExtension) != 0 {
+				return fmt.Errorf("non-commit vote extension present; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+			if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit && len(vote.ExtensionSignature) != 0 {
+				return fmt.Errorf("non-commit vote extension signature present; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+		} else { // vote extensions disabled
+			if len(vote.VoteExtension) != 0 {
+				return fmt.Errorf("vote extension present but extensions disabled; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
```

Comment:
Can this cause an issue to a new validator since he or she would not be expected to sign the first batch?